oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java

```diff
@@ -192,6 +201,61 @@ String createTokenUrlWithScopes() {
     return tokenUrl.toString();
   }
 
+  /**
+   * Gets the universe domain from the GCE metadata server.
+   *
+   * <p>Returns an explicit universe domain if it was provided during credential initialization.
+   *
+   * <p>Returns the {@link Credentials#GOOGLE_DEFAULT_UNIVERSE} if universe domain endpoint is
+   * unavailable or returns an empty string.
+   *
+   * <p>Otherwise, returns universe domain from GCE metadata service.
+   *
+   * <p>Any above value is cached for the credential lifetime.
+   *
+   * @throws IOException if a call to GCE metadata service was unsuccessful. Check if exception
+   *     implements the {@link Retryable} and {@code isRetryable()} will return true if the
+   *     operation may be retried.
+   * @return string representing a universe domain in the format some-domain.xyz
+   */
+  @Override
+  public String getUniverseDomain() throws IOException {
+    if (!isDefaultUniverseDomain()) {
+      return super.getUniverseDomain();
+    }
+
+    synchronized (this) {
+      if (universeDomainFromMetadata != null) {
+        return universeDomainFromMetadata;
+      }
+
+      universeDomainFromMetadata = getUniverseDomainFromMetadata();
+      return universeDomainFromMetadata;
+    }
+  }
+
+  private String getUniverseDomainFromMetadata() throws IOException {
+    HttpResponse response = getMetadataResponse(getUniverseDomainUrl());
+    int statusCode = response.getStatusCode();
+    if (statusCode == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+      return Credentials.GOOGLE_DEFAULT_UNIVERSE;
+    }
+    if (statusCode != HttpStatusCodes.STATUS_CODE_OK) {
+      IOException cause =
+          new IOException(
+              String.format(
+                  "Unexpected Error code %s trying to get universe domain"
+                      + " from Compute Engine metadata for the default service account: %s",
+                  statusCode, response.parseAsString()));
+      throw new GoogleAuthException(true, cause);
+    }
+    String responseString = response.parseAsString();
+    if (responseString.isEmpty()) {
+      return Credentials.GOOGLE_DEFAULT_UNIVERSE;
```

Comment:
When would this happen?

Reply:
There was an intermediate design of MDS where endpoint returns empty string for default universe. It was quickly abandoned, but made it into the wild already :) Now we expect either 404 or a non-empty value, but adding this just in case the older build persists somewhere.

Reply:
https://docs.google.com/document/d/1Fk4XRywIKfnVzKUn3I2HffZWrmxhtnnBXWi0-lU0D2M/edit?resourcekey=0-CDsW8urip79pyXK4MZuETw&disco=AAABAOEGSPE

Reply:
Thanks, maybe put a comment here explaining.

Reply:
Done